src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/snapshot/AttemptSnapshotStep.kt

```diff
@@ -0,0 +1,75 @@
+package com.amazon.opendistroforelasticsearch.indexstatemanagement.step.snapshot
+
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.elasticapi.suspendUntil
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.ManagedIndexMetaData
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.action.SnapshotActionConfig
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.model.managedindexmetadata.StepMetaData
+import com.amazon.opendistroforelasticsearch.indexstatemanagement.step.Step
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequest
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.rest.RestStatus
+
+class AttemptSnapshotStep(
+    val clusterService: ClusterService,
+    val client: Client,
+    val config: SnapshotActionConfig,
+    managedIndexMetaData: ManagedIndexMetaData
+) : Step(name, managedIndexMetaData) {
+
+    private val logger = LogManager.getLogger(javaClass)
+    private var stepStatus = StepStatus.STARTING
+    private var info: Map<String, Any>? = null
+
+    @Suppress("TooGenericExceptionCaught")
+    override suspend fun execute() {
+        try {
+            logger.info("Executing snapshot on ${managedIndexMetaData.index}")
+            val createSnapshotRequest = CreateSnapshotRequest()
+                .indices(managedIndexMetaData.index)
+                .snapshot(config.snapshot)
```

Comment:
Should we add a unique hash to the end of the snapshot name? Since duplicate snapshot names are not allowed this depends on how the user defined their policy and snapshot action.

If it's a linear flow, i.e. hot -> warm -> cold -> delete where delete has 1 snapshot executed then it seems fine.

But, if it's a circular flow where it continuously takes snapshots then they have to use date math expressions in the name. If it executes fast enough though, they could be taking multiple snapshots per day and they need to make sure to use the correct date math expression. Might be easier for the user if we just take the user defined snapshot name and append a unique hash at the end? So they can still reference `<there_concrete_name>*`.

Thoughts?

Reply:
We've added the DateTime to the snapshot name